api/application_test.go

```diff
@@ -49,6 +51,7 @@ func (s *ApplicationSuite) BeforeTest(suiteName, testName string) {
 }
 
 func (s *ApplicationSuite) AfterTest(suiteName, testName string) {
+	auth.UseCryptoRand()
```

Comment:
I'd prefer if this is not exported in the public api, the public api should only provide the crypto rand.

Having something only in the package has serveral benefits:
* less public api which is only used for tests
* no possible wrong usage of the public api (when only crypto is available)

application.go
```go
var authApplicationToken = authGenerateApplicationToken

func (a *ApplicationAPI) CreateApplication(ctx *gin.Context) {
	app := model.Application{}
	if err := ctx.Bind(&app); err == nil {
		app.Token = auth.GenerateNotExistingToken(authGenerateApplicationToken, a.applicationExists)
		app.UserID = auth.GetUserID(ctx)
		app.Internal = false
		a.DB.CreateApplication(&app)
		ctx.JSON(200, withResolvedImage(&app))
	}
}
```
application_test.go
```go
func someTest() {
    old := authApplicationToken
    authApplicationToken = test.Tokens("Abbbbb", "...")
    defer func() {
         authApplicationToken = old
    }
    // do something with the tokens
}
```

Reply:
The problem is the mechanism to switch random sources from default crypto to math Rand must call a API or change a global variable in the auth package, no matter from api or test package. Exporting an API instead of changing global variable enables users and developers to see warnings on misuse of this exported API.

Reply:
I agree but this solution has the benefit that all things are package local so there is no public api which can be used to change the rand. I'd differentiate here changing behavior globally and resetting a private global variable. Also, it is also done this way in other parts of Gotify f.ex. here https://github.com/gotify/server/blob/master/api/stream/client.go#L14

Reply:
So we would need to make a copy of Generate(Image|Application|Client)Token in the API package. I could do that if you really want to, but I think the benefit it brings us is really just about as much as the more code to maintain it brings us:(

Also, maybe I think we are going in the wrong direction, instead of making complicated guesses about the generated token, why not just pull it out from test database and verify if the token is valid?

Reply:
Yeah that would be ok as well. Another option would be to add it to the *Api struct.

Reply:
I would update this pr tomorrow.

Reply:
Okay, thanks.